File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/ScreenCaptureService.java
package jp.gr.java_conf.snake0394.loglook_android;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Service;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.PixelFormat;
import android.hardware.display.DisplayManager;
import android.hardware.display.VirtualDisplay;
import android.media.Image;
import android.media.ImageReader;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.util.DisplayMetrics;
import android.view.WindowManager;

import java.nio.ByteBuffer;

import jp.gr.java_conf.snake0394.loglook_android.logger.Logger;
import jp.gr.java_conf.snake0394.loglook_android.view.activity.AcquireScreenShotPermissionActivity;


/**
 * アプリを起動するランチャー
 */
@TargetApi(Build.VERSION_CODES.LOLLIPOP)
public class ScreenCaptureService extends Service {

    private final Handler handler = new Handler();

    private  static ImageReader mImageReader;
    private  static VirtualDisplay mVirtualDisplay;
    private WindowManager wm;
    private static MediaProjection mediaProjection;
    private static MediaProjectionManager mediaProjectionManager;
    
    private static Intent screenshotPermission = null;
    
    private static int displayWidth;
    private static int displayHeight;
    
    
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        Logger.d("ScreenShotService", "onCreate");

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            stopSelf();
            return;
        }
    
        while (mediaProjection == null) {
            getScreenshotPermission();
        }

        setTheme(R.style.AppTheme);

        wm = (WindowManager) getSystemService(WINDOW_SERVICE);

        DisplayMetrics metrics = getResources().getDisplayMetrics();
        wm.getDefaultDisplay()
          .getMetrics(metrics);
        
        displayWidth = metrics.widthPixels;
        displayHeight = metrics.heightPixels;
        int density = metrics.densityDpi;

        mImageReader = ImageReader.newInstance(displayWidth, displayHeight, PixelFormat.RGBA_8888, 2);

        mVirtualDisplay = mediaProjection.createVirtualDisplay("Capturing Display", displayWidth, displayHeight, density, DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR, mImageReader.getSurface(), null, null);
    }

    @Override
    public int onStartCommand(Intent intent, final int flags, int startId) {
        
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mVirtualDisplay.release();
        mediaProjection.stop();
        Logger.d("ScreenShotService", "onDestroy");
    }

    public static Bitmap getScreenshot() {
        Image image = mImageReader.acquireLatestImage();
        if (image == null) {
            return null;
        }
        Image.Plane[] planes = image.getPlanes();
        ByteBuffer buffer = planes[0].getBuffer();

        int pixelStride = planes[0].getPixelStride();
        int rowStride = planes[0].getRowStride();
        int rowPadding = rowStride - pixelStride * displayWidth;

        Bitmap bitmap = Bitmap.createBitmap(displayWidth + rowPadding / pixelStride, displayHeight, Bitmap.Config.ARGB_8888);
        bitmap.copyPixelsFromBuffer(buffer);
        image.close();

        return bitmap;
    }
    
    protected static void getScreenshotPermission() {
        try {
            if (screenshotPermission != null) {
                if(null != mediaProjection) {
                    mediaProjection.stop();
                    mediaProjection = null;
                }
                mediaProjection = mediaProjectionManager.getMediaProjection(Activity.RESULT_OK, (Intent) screenshotPermission.clone());
            } else {
                openScreenshotPermissionRequester();
            }
        } catch (final RuntimeException ignored) {
            openScreenshotPermissionRequester();
        }
    }
    
    protected static void openScreenshotPermissionRequester(){
        final Intent intent = new Intent(App.getInstance() , AcquireScreenShotPermissionActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
        App.getInstance().startActivity(intent);
    }
    
    public static void setScreenshotPermission(final Intent permissionIntent) {
        screenshotPermission = permissionIntent;
    }
}

File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/ScreenCaptureService.kt
package jp.gr.java_conf.snake0394.loglook_android

import android.annotation.TargetApi
import android.app.Service
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.PixelFormat
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.widget.Toast
import jp.gr.java_conf.snake0394.loglook_android.logger.Logger
import jp.gr.java_conf.snake0394.loglook_android.view.activity.ScreenCaptureActivity

@TargetApi(Build.VERSION_CODES.LOLLIPOP)
class ScreenCaptureService : Service() {

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            stopSelf()
            return
        }

        isRunning = true

        setTheme(R.style.AppTheme)

        mediaProjection = ScreenCaptureActivity.getMediaProjection()
        if (mediaProjection == null) {
            stopSelf()
            handler.post({ Toast.makeText(applicationContext, "起動失敗", Toast.LENGTH_SHORT).show() })
            Logger.d("DeckListCaptureService", "mediaProjection = null")
            return
        }

        val metrics = resources.displayMetrics
        OverlayService.getDefaultDisplay().getMetrics(metrics)
        displayWidth = metrics.widthPixels
        displayHeight = metrics.heightPixels
        val density = metrics.densityDpi

        mImageReader = ImageReader.newInstance(displayWidth, displayHeight, PixelFormat.RGBA_8888, 2)

        mVirtualDisplay = mediaProjection.createVirtualDisplay("Capturing Display", displayWidth, displayHeight, density, DisplayManager.VIRTUAL_DISPLAY_FLAG_AUTO_MIRROR, mImageReader.surface, null, null)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Logger.d(TAG, "onStartCommand")
        return Service.START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()

        Logger.d(TAG, "onDestroy")
        mVirtualDisplay.release()
        mediaProjection.stop()
        isRunning = false
    }

    companion object {
        private val TAG = "ScreenCaptureService"
        private val handler by lazy { Handler() }

        private lateinit var mediaProjection: MediaProjection
        private lateinit var mImageReader: ImageReader
        private lateinit var mVirtualDisplay: VirtualDisplay

        private var displayWidth: Int = 0
        private var displayHeight: Int = 0

        @JvmStatic
        var isRunning = false

        @JvmStatic
        fun getScreenshot(): Bitmap {
            val image = mImageReader.acquireLatestImage()
            val planes = image.planes
            val buffer = planes[0].buffer

            val pixelStride = planes[0].pixelStride
            val rowStride = planes[0].rowStride
            val rowPadding = rowStride - pixelStride * displayWidth

            val bitmap = Bitmap.createBitmap(displayWidth + rowPadding / pixelStride, displayHeight, Bitmap.Config.ARGB_8888)
            bitmap.copyPixelsFromBuffer(buffer)
            image.close()

            return bitmap
        }
    }
}
